### ISSUE TITLE
Create a model pipeline for classifying residential property sales

### ISSUE THREAD

**Issue by dfsnow** [ccao-data/model-sales-val#1]
## Context

The CCAO previously relied on a set of simple heuristics to validate sales used for training its models. Typically, this included things like dropping sales X std deviations away from the area mean, or all sales below a certain price. This approach is extremely limited, as it does not remove non-arms-length sales that result from many common situations, such as:

- A property is torn-down and a new property is built and sold. The new high sale price is erroneously attached to the old characteristics because the CCAO never received permits or did not updated the property chars.
- A property is sold to a family member at a below-market-rate price
- A bank purchases a property and then resells it at a discount

In the summer of 2022, the Mansueto Institute partnered with the CCAO to create a number of additional heuristics that can identify some of these common situations. A summary of their work can be [found on the following slide deck](https://docs.google.com/presentation/d/1AR4UfgJeEikKFBrcwSFIDgDWvpEZPvN1hvAj9bjqdbA/edit#slide=id.p), and the relevant code can be found in this repository.

The main output of Mansueto's project is a new taxonomy for classifying sales (see PPT). This taxonomy corresponds to individual features that get attached to each sale. These features can be used alone or in combination to mark a sale as "valid" or "invalid".

## Goal

The goal of this project is to follow-up on Mansueto's work by creating a supervised classification model using their engineered features. The output of this model is a simple binary, is or is not valid. The model output vector will be uploaded to the system of record (iasWorld) and will replace the (currently empty) validated sale data field. Ultimately, this model's output will determine which sales are included in the larger residential and condominium valuation models.

## Outline

This repo should be structured similarly to the [Residential Valuation Model](https://gitlab.com/ccao-data-science---modeling/models/ccao_res_avm), with pipeline stages structured as separate scripts and the pipeline itself run via [DVC](https://dvc.org/). The pipeline should pull data from S3, attach flags created by Mansueto ([flagging.py](./flagging.py)), train a classification model, then use the trained model to classify unseen sales.

The necessary labels for the training data will be produced by Valuations. We are currently still waiting on this data.

A potential pipeline for this model might be:

1. Train - Create the trained model using the already labeled and featurized data
2. Ingest - Pull raw sales data from S3, use Mansueto's scripts to construct features for each sale. Pull only sales that do not already have labels
3. Predict - Use the trained model to classify the new unseen sales
4. Export - Attach metadata to each sale (when it was classified, model ID, etc.), then export back to S3 and/or prepare for upload to iasWorld (TBD on this)

DVC will cache stages with no changes so that they do not need to be re-run. So this pipeline structure would cache the first stage (since the training data and model specification won't change often), then run the rest of the pipeline. The model object itself will also be cached to S3. Eventually, we should automate this pipeline to run on any new sales pulled from IDOR.

### Getting Started

@spavlekovsky you can get started on the project by getting familiar with the existing structure of the residential modeling pipeline and Mansueto's existing flags. I would advise that you copy a lot of the existing code and repo. Finally, create the first stage of the pipeline outlined above. We're waiting on real training data, but I can give you a fake training set so you can play with the pipeline structure and model.

## Requirements

- Work in the branch attached to this issue, not in the master branch
- The model pipeline should be built in R for the sake of consistency with other models, but feature engineering and other tasks can be done in Python. You can run python directly in R using `reticulate`
- The main output of this project for @spavlekovsky is a trained model object and basic model pipeline that can classify unseen sales. The more complicated part (operationalizing and integrating the model with existing systems) can be left to @dfsnow 

## Outstanding Questions

- How should the data get into iasWorld? Batch update? API?
- How often does the model need to be run? Should we schedule it to run as we get new sales?
- How should we operationalize it? Can we pull new sales from S3 then push to iasWorld, running the model on GitLab CI?
- When should it be retrained? How often will we need to relabel sales?
- What time period of sales should we run it on? All past sales? Only the past few years?
- What if we retrain the model? If the outcome changes, do we update previously flagged records?

**Comment by wagnerlmichael** [comment]
This sales val process has been integrated with our data lake, and we are currently in the process of working towards an iasWorld upload so these flags can live there as well.
